distributed/protocol/core.py

```diff
@@ -122,7 +122,12 @@ def loads(frames, deserialize=True, deserializers=None):
             else:
                 value = Serialized(head, fs)
 
-            get_in(key[:-1], msg)[key[-1]] = value
+            # TODO: This is problematic with tuple assignment
+            val_holder = get_in(key[:-1], msg)
+            if isinstance(val_holder, tuple):
+                msg[key[0]] = list(val_holder)
+                val_holder = get_in(key[:-1], msg)
+            val_holder[key[-1]] = value
```

Comment:
Why this change?  Also, should we be concerned about performance impact here?

Reply:
So this is basically a result of the msgpack stuff no longer returning lists ever. 

Since tuple([1,2])[0] moans about being immutable :(

Reply:
Where do we assume mutability?  Also, does this handle lists of lists of lists?  

I'm tempted to assume immutability and explicitly go mutable when necessary.  I'm running the test suite now.

Reply:
Refactored this to support lists of lists.

Reply:
Refactored this to support deeper nested things.  This does unfortunately mean a recursive `put_in` (similar to `get_in`) needs to be around

Reply:
Sorry, I should have been more clear.  I'm inclined, if possible, to just use immutable data structures everywhere and then see where this breaks and make those specific instances explicitly mutable, by converting to lists at that location.  Do you have thoughts on this approach?

Reply:
Ping on this.  @mariusvniekerk do you think it would work to set `use_lists=False`, use tuples everywhere, and then manually handle the hopefully few cases where we mutate a result?